compiler-core/src/build/project_compiler.rs

```diff
@@ -130,6 +135,33 @@ where
         Ok(Package { config, modules })
     }
 
+    /// Checks that version file found in the build directory matches the
+    /// current version of gleam. If not, we will clear the build directory
+    /// before continuing. This will ensure that upgrading gleam will not leave
+    /// one with confusing or hard to debug states.
+    pub fn check_gleam_version(&self) -> Result<(), Error> {
+        let build_path = paths::build();
+        let version_path = paths::build_version();
+        if self.io.is_file(&version_path) {
+            let version = self.io.read(&version_path)?;
+            if version == self.options.gleam_version {
+                return Ok(())
+            }
+        }
```

Comment:
Could we log a message here?
```suggestion
        }

        tracing::info!("removing_build_state_from_different_gleam_version");

```